Add peer.service semantic convention to indicate the name of a target… (#652)
open-telemetry/opentelemetry-specification: specification/trace/semantic_conventions/rpc.md

#### Service name

> On the server process receiving and handling the remote procedure call, the service name provided in `rpc.service` does not necessarily have to match the [`service.name`][] or [`peer.service`][] resource attributes. One process can expose multiple RPC endpoints and thus have multiple RPC service names. From a deployment perspective, as expressed by the `service.*` resource attributes, it will be treated as one deployed service with one `service.name`.

Reviewer: ```suggestion
On the server process receiving and handling the remote procedure call, the service name provided in `rpc.service` does not necessarily have to match the [`service.name`][] resource attribute.
One process can expose multiple RPC endpoints and thus have multiple RPC service names. From a deployment perspective, as expressed by the `service.*` resource attributes, it will be treated as one deployed service with one `service.name`.
Likewise, on clients sending RPC requests to a server, the service name provided in `rpc.service` does not have to match the [`peer.service`][] span attribute.
```

`peer.service` is not a resource attribute.

As an example, given a process deployed as `QuoteService`, this would be the name that goes into the `service.name` resource attribute which applies to the entire process.
This process could expose two RPC endpoints, one called `CurrencyQuotes` (= `rpc.service`) with a method called `getMeanRate` (= `rpc.method`) and the other endpoint called `StockQuotes`  (= `rpc.service`) with two methods `getCurrentBid` and `getLastClose` (= `rpc.method`).

> Generally, a user SHOULD not set `peer.service` to a fully qualified RPC service name as that will be redundant with `rpc.service`.

Reviewer 2: ```suggestion
Generally, a user SHOULD not set `peer.service` if it would be equal to the RPC service name as that would be redundant with `rpc.service`.
```

Or do we want users to set peer.service to the unqualified name of the RPC service?

Author: Thanks or the suggestions - for this one, I don't think we should encourage anything and leave it to the user. Personally, I find an unqualified service name to be easier to use in monitoring than the fully qualified (fully qualified I use to ensure no code collisions in code, not since it looks nice in dashboards, it often doesn't).

Reviewer 2: I think there is yet another difference between the fully qualified RPC service name (that is used to connect to the service for example) and the fully qualified name of the (Java/Go/...) class that implements the service.

Author: Yeah - but I don't think people generally design their proto package names based on what they want to see in monitoring. In practice, I often see Java shops use a long FQDN proto package, C++ shops use a one-worder or so, and I've seen the package change from a one-worder to a FQDN because code stopped compiling due to a collision. Letting the user decouple this programming concept from monitoring seems nice for them.

Reviewer 3: a) `s/SHOULD not/SHOULD NOT/` (this is how RFC keywords work)
b) I am not sure about this sentence at all. If anything, I would rather people not send `rpc.service`, but do send `peer.service`, as the latter can be used in dependency diagrams, but `rpc.service` is meh (instrumentation at Uber just sets `span.name={rpc.service}::{rpc.method}` and does not send those extra tags at all, there's little use for them).

Author: @arminru felt pretty strongly about adding this line. I'm ok with either though :)

Reviewer 3: I think the sentence emphasizes the wrong thing. I am not bothered by redundancy, because it is coincidental. Earlier we explain that service name is a reference to a deployment unit, like "k8s service". So I agree with the SHOULD NOT part, but because `peer.service` and `rpc.service` are completely different things. The paragraph above this one explains it quite well with the `QuoteService` example.

Ideally we should have this explanation in [resource#service][1].

[1]: https://github.com/open-telemetry/opentelemetry-specification/blob/462b64a0b43bf951d65e3489abecf5f9347e0c06/specification/resource/semantic_conventions/README.md#service

Author: Thanks - didn't tweak resource.md, but I think I get what you're saying. I tweaked to remove the point about redundancy and tie into the above paragraph, hopefully this helps

As an example, given a process deployed as `QuoteService`, this would be the name that goes into the `service.name` resource attribute which applies to the entire process.
This process could expose two RPC endpoints, one called `CurrencyQuotes` (= `rpc.service`) with a method called `getMeanRate` (= `rpc.method`) and the other endpoint called `StockQuotes`  (= `rpc.service`) with two methods `getCurrentBid` and `getLastClose` (= `rpc.method`).

> The same applies for `peer.service`, and generally, a user SHOULD NOT set `peer.service` to a fully qualified RPC service name.

Reviewer: It is unclear what "the same applies" refers to exactly.

```suggestion
In this example, spans representing client request should have their `peer.service` attribute set to `QuoteService` as well to match the server's `service.name` resource attribute.
Generally, a user SHOULD NOT set `peer.service` to a fully qualified RPC service name.
```

Author: Thanks!
